iroh/src/node/protocol.rs

```diff
@@ -0,0 +1,127 @@
+use std::{any::Any, collections::BTreeMap, fmt, sync::Arc};
+
+use anyhow::Result;
+use futures_lite::future::Boxed as BoxedFuture;
+use futures_util::future::join_all;
+use iroh_net::endpoint::Connecting;
+
+use crate::node::DocsEngine;
+
+/// Handler for incoming connections.
+///
+/// An iroh node can accept connections for arbitrary ALPN protocols. By default, the iroh node
+/// only accepts connections for the ALPNs of the core iroh protocols (blobs, gossip, docs).
+///
+/// With this trait, you can handle incoming connections for custom protocols.
+///
+/// Implement this trait on a struct that should handle incoming connections.
+/// The protocol handler must then be registered on the node for an ALPN protocol with
+/// [`crate::node::builder::ProtocolBuilder::accept`].
+pub trait ProtocolHandler: Send + Sync + IntoArcAny + fmt::Debug + 'static {
+    /// Handle an incoming connection.
+    ///
+    /// This runs on a freshly spawned tokio task so this can be long-running.
+    fn accept(self: Arc<Self>, conn: Connecting) -> BoxedFuture<Result<()>>;
+
+    /// Called when the node shuts down.
+    fn shutdown(self: Arc<Self>) -> BoxedFuture<()> {
+        Box::pin(async move {})
+    }
+}
+
+/// Helper trait to facilite casting from `Arc<dyn T>` to `Arc<dyn Any>`.
+///
+/// This trait has a blanket implementation so there is no need to implement this yourself.
+pub trait IntoArcAny {
+    fn into_arc_any(self: Arc<Self>) -> Arc<dyn Any + Send + Sync>;
+}
+
+impl<T: Send + Sync + 'static> IntoArcAny for T {
+    fn into_arc_any(self: Arc<Self>) -> Arc<dyn Any + Send + Sync> {
+        self
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub(super) struct ProtocolMap(BTreeMap<&'static [u8], Arc<dyn ProtocolHandler>>);
+
+impl ProtocolMap {
+    /// Returns the registered protocol handler for an ALPN as a concrete type.
+    pub fn get_typed<P: ProtocolHandler>(&self, alpn: &[u8]) -> Option<Arc<P>> {
```

Comment:
Could we have the same visibility marker here as on the struct, so `pub(super)`?  I find that a lot easier to read.  Plus there's fewer surprises: if the super module decides to export this than suddenly these functions are really pub, and that might have been unintended.